src/common/libsubprocess/subprocess.h

```diff
@@ -73,6 +73,12 @@ enum {
      * error.
      */
     FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF = 8,
+    /* flux_local_exec(): when an on_credit callback is specified, the
+     * full initial credits are sent as the first callback.  However,
+     * in some circumstances this may not be desired.  This flag will
+     * disable the sending of "initial credits".
+     */
+     FLUX_SUBPROCESS_FLAGS_NO_INITIAL_CREDITS = 16,
```

Comment:
As stated in the earlier comment, this flag feels hacky to me.  

Proposal: eliminate it and always send the first `on_credit` callback.

Subprocess users that would have used this flag, including the subprocess server, would track the current available buffer space.  If they want to issue remote credits or write data before the first `on_credit()` callback, they would set the current buffer space negative indicating that some number of credits were borrowed.  The first callback would cause the tracked count to reach zero (no action should be taken because there is no available buffer space).  The next call would result in a positive value.

Reply:
> Subprocess users that would have used this flag, including the subprocess server, would track the current available buffer space. If they want to issue remote credits or write data before the first on_credit() callback, they would set the current buffer space negative indicating that some number of credits were borrowed. The first callback would cause the tracked count to reach zero (no action should be taken because there is no available buffer space). The next call would result in a positive value.

Going into the weeds, I remember why I did this flag.  The `on_credit()` callback's prototype is

`void cb (flux_subprocess_t *p, const char *stream, int bytes)`

So if we wanted to track "current buffer space" (or more simply a flag indicating "did we already get initial credits") we have to use a hash to lookup the stream.  

So I felt it was better to use the flag.

A hash lookup on every `on_credit()` call is certainly not the end of the world.  It is what it is.  But I did dislike adding that data structure / lookup.

One solution to avoid the hash is to remove the server's "initial credits" response which groups all of the channel's initial credits into one RPC response.  Instead, we just forward each channel's "initial credits" from the local subprocess, i.e. multiple RPC responses if there are multiple writable channels.

So it's a series of pros and cons on each of the approaches.  Any strong feelings on any of them?  To summarize

- option 1 (what's currently done): flag for no initial credits from local subprocess, the server sends 1 RPC for all initial channel credits

- option 2 (what I believe you proposed): track "space" for each writable channel, only respond with non-initial credits.  Requires a hash and a lookup on each credit callback from local subprocess.

- option 3 (what I propose above): do not have server collect all writable channel credits in 1 RPC, forward each RPC from local subprocess.  multiple RPC responses, but no hash necessary.

Reply:
My proposal this morning is not option 2, it is:

> What about a compromise where we eliminate the flag and the dedicated watcher for the first callback, and just always send the first on_credit() response?

> So if we wanted to track "current buffer space" (or more simply a flag indicating "did we already get initial credits") we have to use a hash to lookup the stream.

My thought was just that those that want to immediately send some data without waiting for the first on_credit() response would "borrow" credits by setting their tally negative.  Then when the initial callback comes, it is added to the negative tally.

Isn't there already a hash of channels in `struct flux_subprocess`?  Would adding a `channel_lookup()` function help?  And maybe add a channel aux container?

Reply:
> My thought was just that those that want to immediately send some data without waiting for the first on_credit() response would "borrow" credits by setting their tally negative. Then when the initial callback comes, it is added to the negative tally.

Yes, but that would require tracking of the credits on the libsubprocess server side, thus the hash lookup on each on_credit() call.

> Isn't there already a hash of channels in struct flux_subprocess? Would adding a channel_lookup() function help? And maybe add a channel aux container?

Oh duh, you're right, there should already be the channel hash.

I dislike the lookup of the channel on every on_credit() call, but I guess it's not a huge deal in the end.

Reply:
I just realized a subtlety in this conversation, which maybe explains some of comments/responses above.

At one point in time I viewed `FLUX_SUBPROCESS_FLAGS_NO_INITIAL_CREDITS` as a useful flag that users could use and wanted it supported for both local & remote subprocesses.

Now, it is simply a convenience flag used pretty much exclusively by the server when launching its local subprocesses.  

So I'm not sure if that removes some feeling of it being "hackish".  Perhaps if there was a hidden/internal flag that could do the same behavior, it wouldn't be viewed as such a hack?

Reply:
Yes, avoiding unnecessary API complexity was my biggest concern.  Making the libsubprocess internals as understandable as possible for the next person who comes along to fix a bug or whatever is a secondary concern, though.

The hash lookup doesn't seem like a big deal to me.  When the server handles a `rexec.write` RPC, it has to find the channel with a hash lookup.  This is just the same in the reverse direction. If the hash lookup helps the code be more understandable by making the subprocess server behave like any subprocess user implementing flow control would, then I'm for the hash lookup.

Reply:
Uh, I was riding my bike and it occurred to me that your option 3 above would be fine!  For most cases, there is only one writable channel (maybe two), so having an RPC response per writable channel should be fine.

It seems like that would achieve the goal of simplifying the code best of all since the first response wouldn't need to be handled by a prep watcher.  The fbuf write watcher could just be started and an initial POLLOUT event would let the credits be sent the same way credits are sent after that.

Reply:
> Uh, I was riding my bike and it occurred to me that your option 3 above would be fine! For most cases, there is only one writable channel (maybe two), so having an RPC response per writable channel should be fine.

Shall we keep the "channels" object that has the potential for > 1 channel per object?  No longer needed now, but it's always possible we may > 1 channel in the future.

Reply:
Yeah seems like it would be OK to leave it as proposed unless it causes you grief.